code/modules/power/singularity/singularity.dm

```diff
@@ -117,8 +117,18 @@
 	return
 
 
-/obj/singularity/bullet_act(obj/item/projectile/P)
-	qdel(P)
+/obj/singularity/bullet_act(obj/item/projectile/bullet/SRN_rocket/N, obj/item/projectile/P, mob/living/user)
+	var/turf/T = get_turf(src)
+	if(N)
+		for(var/mob/M in GLOB.player_list)
+			if(M.get_virtual_z_level() == get_virtual_z_level())
+				SEND_SOUND(M, 'sound/magic/charge.ogg')
+				to_chat(M, "<span class='boldannounce'>You feel reality distort for a moment...</span>")
+				SEND_SIGNAL(M, COMSIG_ADD_MOOD_EVENT, "delam", /datum/mood_event/delam)
+				shake_camera(M, 15, 3)
+		new/obj/singularity/spatial_rift(T)
+	else
+		qdel(P)
```

Comment:
Shouldn't you run on impact instead?

Reply:
I tried impact but I think singularities consume or delete the projectile before actual impact. And I think it needs to be able to proc onhit for impact(comment from impact proc). But I am not totally sure because I am learning this code on the go.
/**
 * Called when the projectile hits something
 * This can either be from it bumping something,
 * or it passing over a turf/being crossed and scanning that there is infact
 * a valid target it needs to hit.
 * This target isn't however necessarily WHAT it hits
 * that is determined by process_hit and select_target.
 *
 * Furthermore, this proc shouldn't check can_hit_target - this should only be called if can hit target is already checked.
 * Also, we select_target to find what to process_hit first.
 */